File: go/core/agent_test.go
package core

import (
	"context"
	"strings"
	"testing"
)

// fakeClient returns scripted responses and records the requests it saw.
type fakeClient struct {
	scripted []ChatResponse
	calls    []ChatRequest
}

func (f *fakeClient) Chat(_ context.Context, req ChatRequest) (ChatResponse, error) {
	f.calls = append(f.calls, req)
	resp := f.scripted[0]
	f.scripted = f.scripted[1:]
	return resp, nil
}

func TestKnowledgeRanksByOverlap(t *testing.T) {
	kb := &InMemoryKnowledge{}
	kb.Ingest("The return window is 17 days from delivery.", "returns.md")
	kb.Ingest("Gift wrapping costs 4.99 per item.", "wrapping.md")
	hits := kb.Query("what is the return window?", 1)
	if len(hits) != 1 {
		t.Fatalf("want 1 hit, got %d", len(hits))
	}
	if !strings.Contains(hits[0].Content, "17 days") {
		t.Fatalf("top hit should be the returns doc, got %q", hits[0].Content)
	}
}

func TestTextReplyStopsAfterOneCall(t *testing.T) {
	client := &fakeClient{scripted: []ChatResponse{{Content: "the answer is 42"}}}
	agent := NewSmoothAgent(client, AgentOptions{Instructions: "be helpful"})
	res, err := agent.Run(context.Background(), "what is the answer?", nil)
	if err != nil {
		t.Fatal(err)
	}
	if res.Text != "the answer is 42" || res.Iterations != 1 || res.ToolCalls != 0 {
		t.Fatalf("unexpected result: %+v", res)
	}
}

func TestToolCallThenFinish(t *testing.T) {
	echo := FuncTool{
		ToolName: "echo",
		Desc:     "Echoes input back",
		Params:   map[string]any{"type": "object", "properties": map[string]any{"text": map[string]any{"type": "string"}}, "required": []string{"text"}},
		Fn: func(_ context.Context, args map[string]any) (string, error) {
			s, _ := args["text"].(string)
			return s, nil
		},
	}
	client := &fakeClient{scripted: []ChatResponse{
		{ToolCalls: []ToolCall{{ID: "call-1", Name: "echo", Arguments: `{"text": "hello tools"}`}}},
		{Content: "done"},
	}}
	agent := NewSmoothAgent(client, AgentOptions{Tools: []Tool{echo}})
	res, err := agent.Run(context.Background(), "use echo", nil)
	if err != nil {
		t.Fatal(err)
	}
	if res.Text != "done" || res.ToolCalls != 1 {
		t.Fatalf("unexpected result: %+v", res)
	}
	// The tool result was fed back as a tool-role message before the final call.
	second := client.calls[1].Messages
	found := false
	for _, m := range second {
		if m.Role == "tool" && m.Content == "hello tools" {
			found = true
		}
	}
	if !found {
		t.Fatalf("tool result not fed back; messages=%+v", second)
	}
}

func TestKnowledgeInjectedIntoSystemPrompt(t *testing.T) {
	kb := &InMemoryKnowledge{}
	kb.Ingest("The return window is exactly 17 days from delivery.", "returns.md")
	client := &fakeClient{scripted: []ChatResponse{{Content: "17 days"}}}
	agent := NewSmoothAgent(client, AgentOptions{Instructions: "support agent", Knowledge: kb})
	if _, err := agent.Run(context.Background(), "how many days to return?", nil); err != nil {
		t.Fatal(err)
	}
	first := client.calls[0].Messages
	if len(first) == 0 || first[0].Role != "system" || !strings.Contains(first[0].Content, "17 days") {
		t.Fatalf("knowledge not injected into system prompt; first=%+v", first)
	}
}

File: typescript/core/test/agent.test.ts
/**
 * Non-network unit tests for the TypeScript core: the agentic loop, tool calling,
 * and knowledge injection, driven by a fake OpenAI-compatible client. Always
 * green (no credentials) — live-gateway behavior is covered by `evals.test.ts`.
 */

import { describe, expect, it } from 'vitest';
import { AgentOptions, ChatClientLike, SmoothAgent, Tool } from '../src/agent.js';
import { InMemoryKnowledge } from '../src/knowledge.js';

type ScriptedMessage = {
    content: string | null;
    tool_calls?: Array<{ id: string; function: { name: string; arguments: string } }> | null;
};

class FakeClient implements ChatClientLike {
    readonly calls: Array<Record<string, unknown>> = [];
    private readonly scripted: ScriptedMessage[];

    constructor(scripted: ScriptedMessage[]) {
        this.scripted = [...scripted];
    }

    chat = {
        completions: {
            create: async (body: Record<string, unknown>) => {
                this.calls.push(body);
                const message = this.scripted.shift()!;
                return { choices: [{ message }] };
            },
        },
    };
}

function makeAgent(client: ChatClientLike, options: AgentOptions = {}): SmoothAgent {
    return new SmoothAgent(client, options);
}

describe('InMemoryKnowledge', () => {
    it('ranks by token overlap', () => {
        const kb = new InMemoryKnowledge();
        kb.ingest('The return window is 17 days from delivery.', 'returns.md');
        kb.ingest('Gift wrapping costs 4.99 per item.', 'wrapping.md');
        const hits = kb.query('what is the return window?', 1);
        expect(hits).toHaveLength(1);
        expect(hits[0].content).toContain('17 days');
    });
});

describe('SmoothAgent', () => {
    it('stops after one call on a text reply', async () => {
        const client = new FakeClient([{ content: 'the answer is 42' }]);
        const agent = makeAgent(client, { instructions: 'be helpful' });
        const result = await agent.run('what is the answer?');
        expect(result.text).toBe('the answer is 42');
        expect(result.iterations).toBe(1);
        expect(result.toolCalls).toBe(0);
    });

    it('runs a tool then finishes', async () => {
        const echo: Tool = {
            name: 'echo',
            description: 'Echoes input back',
            parameters: { type: 'object', properties: { text: { type: 'string' } }, required: ['text'] },
            execute: async (args) => String(args.text ?? ''),
        };
        const client = new FakeClient([
            { content: null, tool_calls: [{ id: 'call-1', function: { name: 'echo', arguments: '{"text": "hello tools"}' } }] },
            { content: 'done' },
        ]);
        const agent = makeAgent(client, { tools: [echo] });
        const result = await agent.run('use echo');
        expect(result.text).toBe('done');
        expect(result.toolCalls).toBe(1);
        // The tool result was fed back as a tool-role message before the final call.
        const secondCallMessages = client.calls[1].messages as Array<Record<string, unknown>>;
        expect(secondCallMessages.some((m) => m.role === 'tool' && m.content === 'hello tools')).toBe(true);
    });

    it('injects knowledge into the system prompt', async () => {
        const kb = new InMemoryKnowledge();
        kb.ingest('The return window is exactly 17 days from delivery.', 'returns.md');
        const client = new FakeClient([{ content: '17 days' }]);
        const agent = makeAgent(client, { instructions: 'support agent', knowledge: kb });
        await agent.run('how many days to return?');
        const messages = client.calls[0].messages as Array<Record<string, unknown>>;
        expect(messages[0].role).toBe('system');
        expect(messages[0].content).toContain('17 days');
    });
});
